### PR TITLE
Propagate previously unused `NOTIFICATION_WORLD_2D_CHANGED`, make CanvasItem/CollisionObject2D use it

### LINKED ISSUE #52423: Separate Viewport/World2D with another viewport sharing the World2D causes physics errors, no rendering
### Godot version

3.3.3-stable_mono_win64

### System information

Windows 10 (64-bit) , GLES 3, Nvidia RTX 3080

### Issue description

I'm using separate `Viewport`s and `World2D`s to have separate locations with their own physics and such. So what I'm doing is having a node containing each location as a `Viewport` with their own `World2D`, and then for the "player" viewport (the one currently rendering) I set it to use the same World2D. (Optionally with a `Camera2D` being moved using `RemoteTransform2D`, but that wasn't necessary to demonstrate the issue in the test project.) This causes some things not to render (MeshInstance2D at least; in my actual project one of the sprites is rendering, but not the other.), and this error to be repeated:

```
ERROR: body_test_motion: Condition "!body->get_space()" is true. Returned: false
   At: servers/physics_2d/physics_2d_server_sw.cpp:1046
ERROR: body_test_ray_separation: Condition "!body->get_space()" is true. Returned: false
   At: servers/physics_2d/physics_2d_server_sw.cpp:1058
```

I've confirmed that the viewports and their children all have the same `World2D` instance being used with `print` and `str`.

I've used the same setup before, except in 3D, where it worked fine. In the minimal reproduction project, I included a 3D version, which works fine.

### Steps to reproduce

The minimal reproduction project has everything set up already, with the startup scene that causes problems, `PlayScene2D`. You can change the startup scene to `PlayScene3D` to see the 3D version that works fine.

### Minimal reproduction project

[LocationTest.zip](https://github.com/godotengine/godot/files/7112161/LocationTest.zip)

### PR DESCRIPTION
**Pull request by spacechase0**
(Master version of #52761)

I took into account the feedback on the previous PR (and put one comment there, since a suggested change didn't work).

New test project (converted one from #52423): [LocationTest.zip](https://github.com/godotengine/godot/files/7930291/LocationTest.zip) (One small difference is I had to change the default environment background color for the 3D test so you could actually see things, since for some reason with master it doesn't render correctly, and everything is black instead. Not sure what the deal with that is.)

*Bugsquad edit:*
- Fixes #52423

### CODE REVIEW

**Comment by reduz** [comment]
So, this is actually really good and needed, the problem is that in 3D you have ENTER_WORLD and EXIT_WORLD notifications and this PR would make 2D work inconsistently with that.

That said, I think the functionality in this PR is better, and changing the 3D notifications to a single NOTIFICATION_WORLD_3D_CHANGED would be more fit, as in most cases you could just use enter/exit tree and handle the 3D world change separately.

So,would you be up to do this change on the 3D side too to make it remain consistent?

**Comment by spacechase0** [comment]
Sorry for the huge delay in response on this - is changing 3D to match still something that is desired since Godot 4 is in beta now? Or should the 2D functionality here be made to match? Or neither?

File: scene/main/viewport.cpp [hunk 1]
```diff
@@ -1001,7 +1001,14 @@ void Viewport::set_world_2d(const Ref<World2D> &p_world_2d) {
 	}
 
 	if (p_world_2d.is_valid()) {
+		bool do_propogate = true;
+		if (world_2d.is_valid() && is_inside_tree()) {
+			do_propogate = true;
+		}
```
Comment (KoBeWi) [inline scene/main/viewport.cpp:1007]: `do_propogate` is always true. I think it should be
```suggestion
		bool do_propogate = world_2d.is_valid() && is_inside_tree();
```

File: scene/main/viewport.cpp [hunk 2]
```diff
@@ -3757,6 +3764,29 @@ float Viewport::get_texture_mipmap_bias() const {
 
 #endif // _3D_DISABLED
 
+void Viewport::_propogate_world_2d_changed(Node *p_node) {
+	if (p_node != this) {
+		if (!p_node->is_inside_tree()) { // may not have entered scene yet
```
Comment (KoBeWi) [inline scene/main/viewport.cpp:3769]:
```suggestion
		if (!p_node->is_inside_tree()) { // May not have entered scene yet.
```

File: scene/main/viewport.h [hunk 3]
```diff
@@ -725,6 +725,8 @@ class Viewport : public Node {
 	bool is_using_xr();
 #endif // _3D_DISABLED
 
+	void _propogate_world_2d_changed(Node *p_node);
```
Comment (KoBeWi) [inline scene/main/viewport.h:728]:
```suggestion
	void _propagate_world_2d_changed(Node *p_node);
```

**Comment by KoBeWi** [comment]
IMO the PR is fine in the current form. It's too late to change the 3D side, so we'll have to live with inconsistency I guess.

File: scene/main/viewport.cpp (hunk 2 above)
Comment (spacechase0) [reply scene/main/viewport.cpp:3766]: Do you want me to remove the condition entirely then?

**Comment by spacechase0** [comment]
Should I squash the PR into 1 commit as well? Or at least the last 3 commits (since they just say "Update <file>", since I forgot to change them before committing in the web interface).

**Comment by KoBeWi** [comment]
Squash everything into 1 commit.

**Comment by spacechase0** [comment]
Okay, should be good to go.

File: scene/main/viewport.cpp [hunk 4]
```diff
@@ -3807,6 +3811,25 @@ float Viewport::get_texture_mipmap_bias() const {
...
 #endif // _3D_DISABLED
 
+void Viewport::_propagate_world_2d_changed(Node *p_node) {
+	if (p_node != this) {
+		if (Object::cast_to<CanvasItem>(p_node)) {
+			p_node->notification(CanvasItem::NOTIFICATION_WORLD_2D_CHANGED);
+		} else {
+			Viewport *v = Object::cast_to<Viewport>(p_node);
+			if (v) {
+				if (v->world_2d.is_valid()) {
```
Comment (KoBeWi) [inline scene/main/viewport.cpp:3821]:
```suggestion
			if (v && v->world_2d.is_valid()) {
```
Can be simplified.

**Review by KoBeWi — APPROVED** [review]

File: scene/main/viewport.cpp (hunk 4 above)
Comment (akien-mga) [reply scene/main/viewport.cpp:3821]: I guess this was written like this to be consistent with 3D:
```cpp
Viewport *v = Object::cast_to<Viewport>(p_node);
if (v) {
        if (v->world_3d.is_valid() || v->own_world_3d.is_valid()) {
                return;
        }
}
```
Which would need to become `if (v && (v->world_3d.is_valid() || v->own_world_3d.is_valid()))` too.

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**